### PR TITLE
Add more /proc/sys/kernel compatibility files

### PR DESCRIPTION
**Pull request by bet4it**
## Summary

Add several `/proc/sys/kernel` compatibility files used by Linux-oriented tests and applications:

- `/proc/sys/kernel/hostname`
- `/proc/sys/kernel/domainname`
- `/proc/sys/kernel/osrelease`
- `/proc/sys/kernel/version`
- `/proc/sys/kernel/tainted`

The UTS-related procfs files are backed by the current UTS namespace. Writes to `hostname` and `domainname` share the same permission checks and update path as the corresponding syscalls.

`/proc/sys/kernel/tainted` implements a minimal Linux-compatible writable state:

- reads return the current taint mask
- writes require `CAP_SYS_ADMIN`
- writes OR known taint bits into the current mask
- unknown taint bits are ignored
- writing `0` does not clear existing taint bits

## Tests

- Added regression coverage for `/proc/sys/kernel` UTS files.
- Added regression coverage for `/proc/sys/kernel/tainted` OR semantics, unknown-bit filtering, and `CAP_SYS_ADMIN` enforcement.
- Enabled related gVisor proc tests by removing matching blocklist entries.

### CODE REVIEW

File: kernel/src/net/uts_ns.rs [hunk 1]
```diff
@@ -73,18 +73,17 @@ impl UtsNamespace {
...
+    }
 
-        let new_host_name = copy_uts_field_from_user(addr, len as _, ctx)?;
-        debug!(
-            "set host name: {:?}",
-            CStr::from_bytes_until_nul(new_host_name.as_bytes()).unwrap()
-        );
-        self.uts_name.write().nodename = new_host_name;
+    /// Sets a new hostname from kernel-owned bytes.
+    pub fn set_hostname_bytes(&self, bytes: &[u8], posix_thread: &PosixThread) -> Result<()> {
```
Comment (lrh2000) [inline kernel/src/net/uts_ns.rs]: Don't introduce both `set_hostname` and `set_hostname_bytes`. I believe it's unnecessary.

Just fix the existing method's signature.
```patch
-    pub fn set_hostname(&self, addr: Vaddr, len: usize, ctx: &Context) -> Result<()> {
+    pub fn set_hostname(&self, reader: VmReader, posix_thread: &PosixThread) -> Result<()> {
```

File: kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+impl VersionFileOps {
+    pub fn new_inode(parent: Weak<dyn Inode>) -> Arc<dyn Inode> {
+        ProcFile::new(Self, parent, mkmod!(a+r))
+    }
+}
+
+impl ProcFileOps for VersionFileOps {
+    fn read_at(&self, offset: usize, writer: &mut VmWriter) -> Result<usize> {
+        read_uts_name(offset, writer, UtsName::version_bytes)
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs]: No need to do this. Check the existing `VmPrinter` API.
```rust
writeln!(printer, "{}", UtsName::VERSION);
```

File: kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+fn current_uts_namespace() -> Arc<UtsNamespace> {
+    let Some(current_task) = Task::current() else {
+        return UtsNamespace::get_init_singleton().clone();
+    };
+    let Some(posix_thread) = current_task.as_posix_thread() else {
+        return UtsNamespace::get_init_singleton().clone();
+    };
+    let ns_proxy = posix_thread.ns_proxy().lock();
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs]: Access the current thread's NS proxy via thread local. Also, there is no need to check and fall back. There is no need to clone the UTS namespace.
```rust
let current_task = Task::current().unwrap();
let thread_local = current_task.as_thread_local().unwrap();
let ns_proxy = thread_local.borrow_ns_proxy();

writeln!(printer, "{}", ns_proxy.uts_ns().name());
```

This method may need to be removed as well.

File: kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+fn write_uts_name(
+    reader: &mut VmReader,
+    set_value: impl FnOnce(&UtsNamespace, &[u8], &PosixThread) -> Result<()>,
+) -> Result<usize> {
+    let Some(current_task) = Task::current() else {
+        return_errno_with_message!(Errno::EINVAL, "no current task");
+    };
+    let Some(posix_thread) = current_task.as_posix_thread() else {
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs]: No need to check if the current task is the POSIX thread.

File: kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        return_errno_with_message!(Errno::EINVAL, "no current POSIX thread");
+    };
+
+    let requested_len = reader.remain();
+    if requested_len > UTS_PROC_WRITE_BUF_LEN {
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs]: Don't repeat the code. I believe this logic should be removed. We can use `set_domainname(reader: VmReader)` instead. The logic should live in the UTS module.

File: kernel/src/fs/fs_impls/procfs/sys/kernel/tainted.rs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        .to_str()
+        .ok()
+        .map(|str| str.trim())
+        .and_then(|str| str.parse::<u64>().ok())
+        .ok_or_else(|| Error::with_message(Errno::EINVAL, "the value is not a valid integer"))?;
+
+    Ok((val, read_bytes))
+}
+
+fn check_cap_sys_admin() -> Result<()> {
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/tainted.rs]: This method does not need to exist.
```rust
    lsm_hooks::on_capable(lsm_hooks::CapableContext::new(
        UserNamespace::get_init_singleton().as_ref(),
        current_thread!().as_posix_thread().unwrap(),
        CapSet::SYS_ADMIN,
    ))
```

Move the code to the caller.

File: kernel/src/fs/fs_impls/procfs/sys/kernel/tainted.rs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+        let (mask, read_bytes) = read_u64_from(reader)?;
+
+        TAINTED_MASK.fetch_or(mask & TAINT_FLAGS_MASK, Ordering::Relaxed);
+
+        Ok(read_bytes)
+    }
+}
+
+fn read_u64_from(reader: &mut VmReader) -> Result<(u64, usize)> {
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/tainted.rs]: Move the method near `read_u32_from`.

File: kernel/src/fs/fs_impls/procfs/sys/kernel/tainted.rs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+const TAINT_FLAGS_MASK: u64 = (1 << TAINT_FLAGS_COUNT) - 1;
+
+static TAINTED_MASK: AtomicU64 = AtomicU64::new(0);
+
+/// Represents the inode at `/proc/sys/kernel/tainted`.
+pub struct TaintedFileOps;
+
+impl TaintedFileOps {
+    pub fn new_inode(parent: Weak<dyn Inode>) -> Arc<dyn Inode> {
+        // Reference: <https://elixir.bootlin.com/linux/v6.16.5/source/kernel/panic.c#L501>
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/tainted.rs]: Check LLM-generated links. This should be a reference link to the permission mode.
```
https://elixir.bootlin.com/linux/v6.16.5/source/kernel/sysctl.c#L1588
```

File: kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+};
+
+const UTS_PROC_WRITE_BUF_LEN: usize = UtsName::MAX_FIELD_BYTES + 1;
+
+/// Represents the inode at `/proc/sys/kernel/hostname`.
+pub struct HostnameFileOps;
+
+impl HostnameFileOps {
+    pub fn new_inode(parent: Weak<dyn Inode>) -> Arc<dyn Inode> {
+        ProcFile::new(Self, parent, mkmod!(a+r, u+w))
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs:23]: Same for `FileOps` in this file. Add permission link reference.

File: test/initramfs/src/regression/fs/procfs/proc_sys_kernel.c [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+	struct utsname updated;
+
+	CHECK(uname(&original));
+
+	TEST_RES(write_file("/proc/sys/kernel/hostname", "proc-host\n"),
+		 _ret == (ssize_t)strlen("proc-host\n"));
+	TEST_SUCC(uname(&updated));
+	TEST_RES(strcmp(updated.nodename, "proc-host"), _ret == 0);
+
+	set_hostname_checked("syscall-host");
```
Comment (lrh2000) [inline test/initramfs/src/regression/fs/procfs/proc_sys_kernel.c]: Why `checked`? It does not seem necessary as it is wrapper for a single system call. Remove `CHECKED` and use `TEST_SUCC`?

File: test/initramfs/src/regression/fs/procfs/proc_sys_kernel.c [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+		 _ret == 0);
+}
+END_TEST()
+
+FN_TEST(proc_sys_kernel_uts_files_update_namespace)
+{
+	struct utsname original;
+	struct utsname updated;
+
+	CHECK(uname(&original));
```
Comment (lrh2000) [inline test/initramfs/src/regression/fs/procfs/proc_sys_kernel.c]: Just use `TEST_SUCC`.

File: kernel/src/net/uts_ns.rs [hunk 12]
```diff
@@ -149,6 +188,9 @@ pub struct UtsName {
 }
 
 impl UtsName {
+    /// The maximum byte length of a UTS field, excluding the trailing nul.
+    pub const MAX_FIELD_BYTES: usize = UTS_FIELD_LEN - 1;
```
Comment (lrh2000) [inline kernel/src/net/uts_ns.rs]: Revert unnecessary changes in this file after you make changes following the previous suggestions. I believe many changes will become unnecessary. Check them yourselves.

File: kernel/src/fs/fs_impls/procfs/sys/kernel/tainted.rs [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        file::mkmod,
+        procfs::template::{ProcFile, ProcFileOps},
+        vfs::inode::Inode,
+    },
+    prelude::*,
+    process::{UserNamespace, credentials::capabilities::CapSet, posix_thread::AsPosixThread},
+    security::lsm::hooks as lsm_hooks,
+};
+
+// Keep this in sync with Linux's TAINT_FLAGS_COUNT.
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/tainted.rs]:
```suggestion
// Reference: <https://elixir.bootlin.com/linux/v6.16.5/...>
```

Fill the link.

File: test/initramfs/src/regression/fs/procfs/proc_sys_kernel.c [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include "../../common/capability.h"
+#include "../../common/test.h"
+
+#define ARRAY_SIZE(array) (sizeof(array) / sizeof((array)[0]))
+#define UTS_FIELD_LEN 65
+
+static void read_file(const char *path, char *buf, size_t buf_len)
```
Comment (lrh2000) [inline test/initramfs/src/regression/fs/procfs/proc_sys_kernel.c]: It is inconsistent that `read_file` is using `CHECK` but `write_file` is not. Can we make the code more consistent?

**Review by lrh2000 — COMMENTED** [review]
Thanks for adding these procfs files.

File: test/initramfs/src/regression/fs/procfs/proc_sys_kernel.c (hunk 14 above)
Comment (bet4it) [reply test/initramfs/src/regression/fs/procfs/proc_sys_kernel.c]: I kept `write_file` unchecked because some tests intentionally exercise write failures, for example the `EPERM` case after dropping `CAP_SYS_ADMIN`. If `write_file` used `CHECK`, those negative cases would fail inside the helper before the test could verify the expected errno.

To make the difference explicit, I renamed the read helper to `read_file_checked`, since all current read sites expect success and should fail immediately on open/read/close errors.

Does this address your concern, or would you prefer that I also add a `write_file_checked` helper for the success-only write cases and keep `write_file` only for expected failures?

File: kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+    let current_task = Task::current().unwrap();
+    let thread_local = current_task.as_thread_local().unwrap();
+    let ns_proxy = thread_local.borrow_ns_proxy();
+    let uts_name = ns_proxy.unwrap().uts_ns().uts_name();
+    let value = get_value(&uts_name);
+    let mut output = [0u8; UtsName::MAX_FIELD_BYTES + 1];
+    output[..value.len()].copy_from_slice(value);
+    output[value.len()] = b'\n';
+
+    read_uts_bytes(offset, writer, &output[..value.len() + 1])
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs]: Make `VmPrinter::write_bytes` as public. Use it instead.
```ryst
printer.write_bytes(bytes);
printer.write_bytes(b"\n");
```

We should replace `read_uts_string` with:
```patch
-fn read_uts_string(offset: usize, writer: &mut VmWriter, value: &str) -> Result<usize> {
+fn read_uts_bytes(offset: usize, writer: &mut VmWriter, bytes: &[u8]) -> Result<usize> {
```

File: kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+    let posix_thread = current_thread.as_posix_thread().unwrap();
+    let written_len = reader.remain();
+    let value_reader = trim_trailing_newline(reader)?;
+
+    set_value(ns_proxy.unwrap().uts_ns(), value_reader, posix_thread)?;
+
+    Ok(written_len)
+}
+
+fn trim_trailing_newline<'a>(reader: &VmReader<'a>) -> Result<VmReader<'a>> {
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs]: OK. It seems that this is something that may invalidate my previous API design...

I checked the Linux behavior, it seems that it does not only remove the only newline.
```
root@localhost:~/asterinas# echo -e 'localhost\n123' > /proc/sys/kernel/hostname 
root@localhost:~/asterinas# cat /proc/sys/kernel/hostname 
localhost
```

Also there is no "the UTS name is too long" error in this case:
```
root@localhost:~/asterinas# (python3 -c 'print("a" * 1000)') > /proc/sys/kernel/hostname 
root@localhost:~/asterinas# cat /proc/sys/kernel/hostname 
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```

If it's the case, the original API design may be more correct (but see below). But the original "the UTS name is too long" error in the procfs file handling is not very correct so it should still be removed.

I wonder if we should just do this:
```rust
// procfs

fn write_uts() {
    let bytes = [0; UTS_FIELD_LEN];
    todo!("read bytes");
    // This seems to be procfs-specific, keep it in the procs code?
    if let Some(pos) = bytes.find('\n') { bytes[pos] = '\0'; } 
    uts.set_hostname_bytes(bytes); // This will ensure that there is a `\0`.
}

// uts

impl UtsNamespace {
    fn set_hostname_bytes(&self, bytes: [u8; UTS_FIELD_LEN]) {
        if let Some(pos) = bytes.find(0) {
            bytes[pos..].fill(0);
        } else {
             bytes[UTS_FIELD_LEN - 1] = 0;
        }
        self.set_hostname_field(byte);
    }
}
```

Or do you have better approaches?

If we do so, the changes related to the `set_hostname` method may need to be reverted... Sorry for the noise.

---

Also let's add some tests for the two corner cases that I've shown you above. Maybe you can figure out more, like "hello\0world"...

File: kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+
+fn write_current_uts_name(
+    reader: &mut VmReader,
+    set_value: impl FnOnce(&UtsNamespace, VmReader, &PosixThread) -> Result<()>,
+) -> Result<usize> {
+    let current_task = Task::current().unwrap();
+    let thread_local = current_task.as_thread_local().unwrap();
+    let ns_proxy = thread_local.borrow_ns_proxy();
+    let current_thread = current_thread!();
+    let posix_thread = current_thread.as_posix_thread().unwrap();
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs]: I guess these methods (this method and `read_current_uts_name`) may need to be refactored signficiantly due to the previous comments.

Nevertheless, try to insert some blank lines and find the easiest way to get all things that you need to access.

By this, I mean that the existing code can be simplified to something like:
```rust
    let current_task = Task::current().unwrap();
    let posix_thread = current_task.as_posix_thread();

    let thread_local = current_task.as_thread_local().unwrap();
    let ns_proxy = thread_local.borrow_ns_proxy();

    let bytes = ...;
```

If you start with `Task::current()` for some reason, Then, don't use convenient macros like `current_thread!()`. This macro is only convenient if you don't want to start with `Task::current()`. You can call `as_thread()`/`as_posix_thread()` directly on `Task::current()`.

File: test/initramfs/src/regression/fs/procfs/proc_sys_kernel.c (hunk 14 above)
Comment (lrh2000) [reply test/initramfs/src/regression/fs/procfs/proc_sys_kernel.c]: > To make the difference explicit, I renamed the read helper to `read_file_checked`, since all current read sites expect success and should fail immediately on open/read/close errors.

OK. I think this is fine. Thanks!

**Review by lrh2000 — COMMENTED** [review]
Thanks for your revision. To be honest, I did miss some corner cases when I checked the previous version, which now led to design problems. Sorry about that. (Although I think you missed them too)

File: kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs (hunk 16 above)
Comment (bet4it) [reply kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs]: Thanks for checking this. I agree that procfs should not reuse the strict syscall parsing here.

I updated the patch in this direction:

- `sethostname(2)` / `setdomainname(2)` still keep the strict Linux syscall behavior:  `len > 64` returns `EINVAL`.
- `/proc/sys/kernel/{hostname,domainname}` now has procfs-specific parsing: it starts from the current UTS field, applies the write offset, copies at most 64 bytes, stops at the first `'\n'` or `'\0'`, nul-terminates the result, and returns the original requested write length on success. So long procfs writes are truncated instead of failing.
- `UtsNamespace::{set_hostname,set_domainname}` now take an already parsed `[u8; UTS_FIELD_LEN]` and only do permission checking plus normalization of the nul-terminated field. This keeps the syscall-specific and procfs-specific parsing outside of the namespace setter.

So this is close to your suggested split: procfs owns the procfs-specific newline/NUL/truncation logic, while UTS namespace owns normalization/storage.

I also added regression coverage for both `/proc/sys/kernel/hostname` and
`/proc/sys/kernel/domainname`:

- `localhost\n123` becomes `localhost`
- 1000 `a`s become 64 `a`s
- `hello\0world` becomes `hello`

File: kernel/src/net/uts_ns.rs [hunk 18]
```diff
@@ -72,69 +72,85 @@ impl UtsNamespace {
...
 
-fn copy_uts_field_from_user(addr: Vaddr, len: u32, ctx: &Context) -> Result<[u8; UTS_FIELD_LEN]> {
-    if len.cast_signed() < 0 {
-        return_errno_with_message!(Errno::EINVAL, "the buffer length cannot be negative");
-    }
-
-    let user_space = ctx.user_space();
-    let mut reader = user_space.reader(addr, len as usize)?;
-
+pub(crate) fn copy_uts_field_from_reader(reader: &mut VmReader) -> Result<[u8; UTS_FIELD_LEN]> {
```
Comment (lrh2000) [inline kernel/src/net/uts_ns.rs]: I think you should have been told several times (in your multiple PRs) to avoid using the `pub(crate)` visibility marker. I'm not sure if the reviewer presented you with the exact rule, but if not, here it is:
https://github.com/asterinas/asterinas/blob/bf8e15eff0877954deba68c4bf188f700d501572/book/src/to-contribute/coding-guidelines/for-maintainability/rust-specific/crates-and-modules.md?plain=1#L59-L61

So please avoid problems like this in the future.

However, you don't need to address this problem now. I am leaving this single comment for you to read. This PR still has some API design issues and some other problems. I will leave more detailed comments when I can find out more time.

Comment (bet4it) [reply kernel/src/net/uts_ns.rs]: Sorry about this. `pub(crate)` has appeared in several of my previous changes, and I should have been more careful here as well.

No one had pointed me to the exact rule before, but I did have a rough understanding that `pub` is generally preferred over `pub(crate)` in this codebase. In this patch, especially, a previously `pub` item became `pub(crate)`, and I should have noticed that before pushing.

Thanks for pointing me to the exact guideline. I will follow it more carefully in future changes.

File: kernel/src/net/uts_ns.rs [hunk 19]
```diff
@@ -206,6 +228,24 @@ impl UtsName {
...
         }
     };
+
+    /// Returns the hostname bytes.
+    pub fn nodename_bytes(&self) -> &[u8] {
+        uts_field_as_bytes(&self.nodename)
+    }
+
+    /// Returns the NIS domain name bytes.
+    pub fn domainname_bytes(&self) -> &[u8] {
```
Comment (lrh2000) [inline kernel/src/net/uts_ns.rs]: I think we can try this:
```rust
struct UtsName {
    ...: UtsField,
}

/// ...
///
/// Although this is a POD type, tt has a type invariant that requires a nul byte to be present, and
/// that all bytes after the nul byte are also nul bytes. Users outside this module are not allowed
/// to arbitrarily mutate the content.
struct UtsField([u8; Self::MAX_BYTES_WITH_NUL]);

impl UtsField {
    const MAX_BYTES;
    const MAX_BYTES_WITH_NUL;

    fn from_bytes_until_nul() -> Self {}
    fn read_from() -> Result<Self> {}
    fn to_cstr() -> &CStr {}
    fn to_array() -> &[u8; Self::MAX_BYTES_WITH_NUL] {}
}
```

(Ideally `UtsField` should not have `#[derive(Pod)]`, but currently I don't have an elegant solution. I leave the comment as shown in the code snippet.)

So we shouldn't write something like:
```rust
    let mut value = [0u8; UtsName::MAX_FIELD_BYTES];
    let value_len = {
        let uts_name = ns_proxy.unwrap().uts_ns().uts_name();
        let current_value = get_value(&uts_name);
        value[..current_value.len()].copy_from_slice(current_value);
        current_value.len()
    };

    read_uts_bytes(offset, writer, &value[..value_len])
```

Just do this:
```rust
    let current_task = Task::current().unwrap();
    let thread_local = current_task.as_thread_local().unwrap();
    let ns_proxy = thread_local.borrow_ns_proxy();

    let value = {
        let uts_name = ns_proxy.unwrap().uts_ns().uts_name();
        *get_value(&uts_name)
    };

    read_uts_bytes(offset, writer, value.to_cstr().to_bytes())
```

---

I expect `uts_field_as_bytes`/`copy_uts_field_from_reader` should go away. Prefer the associated method like `UtsField::read_from`/`UtsField::to_cstr` instead. The global constant `UTS_FIELD_LEN` should also go away.

File: kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+        new_value[offset] = 0;
+    }
+
+    set_value(uts_ns, new_value, posix_thread)?;
+
+    if copied_len < copy_len {
+        return Ok(copied_len);
+    }
+
+    Ok(requested_len)
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs]: I think this code is too complex. The following version is shorter (even with some comments):
```rust
    let len = reader.remain();
    if offset >= UtsField::MAX_BYTES || len == 0 {
        // Overflowed bytes are consumed but ignored.
        reader.skip(len);
        return Ok(len);
    }

    let current_task = Task::current().unwrap();
    let posix_thread = current_task.as_posix_thread().unwrap();

    let thread_local = current_task.as_thread_local().unwrap();
    let ns_proxy = thread_local.borrow_ns_proxy();
    let uts_ns = ns_proxy.unwrap().uts_ns();

    // We need to edit the old name. Note that we release the lock in between, so there will be
    // some races. However, this should be fine, given that Linux does the same thing, see
    // <https://elixir.bootlin.com/linux/v7.1.2/source/kernel/utsname_sysctl.c#L57-L59>.
    let mut value = if offset != 0 {
        let uts_name = uts_ns.uts_name();
        let current_value = get_value(&uts_name);
        *current_value.to_array()
    } else {
        [0u8; UtsField::MAX_BYTES_WITH_NUL]
    };

    let mut writer = VmWriter::from(&mut value[offset..]).to_fallible();
    let copied_len = reader.read_fallible(&mut writer)?;

    // Truncate any bytes after the newline encountered. Any additional bytes will also be zeroed in
    // `UtsField::from_bytes_until_nul`.
    for byte in &mut value[offset..offset + copied_len] {
        if *byte == '\n' {
            *byte = '\0';
            break;
        }
    }

    let new_value = UtsField::from_bytes_until_nul(&value);
    set_value(uts_ns, new_value, posix_thread)?;

    // Overflowed bytes are consumed but ignored.
    reader.skip(reader.remain());
    Ok(len)
```

---

Also note that procfs does not seem to allow partial write. So I don't think we need to address the `Err((_err, copied_len)) => copied_len,` case. Just use `?` to make the code fail.

File: kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+    printer.write_bytes(b"\n")?;
+
+    Ok(printer.bytes_written())
+}
+
+fn write_current_uts_name(
+    offset: usize,
+    reader: &mut VmReader,
+    get_value: impl FnOnce(&UtsName) -> &[u8],
+    set_value: impl FnOnce(&UtsNamespace, [u8; UtsName::FIELD_BYTES], &PosixThread) -> Result<()>,
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs]: I think `set_value`/`get_value` can be function pointers: `fn(&UtsName) -> &[u8]`.

File: kernel/src/syscall/setdomainname.rs [hunk 22]
```diff
@@ -1,10 +1,23 @@
...
+    syscall::SyscallReturn,
+};
 
 pub fn sys_setdomainname(addr: Vaddr, len: usize, ctx: &Context) -> Result<SyscallReturn> {
     let ns_proxy_ref = ctx.thread_local.borrow_ns_proxy();
     let ns_proxy = ns_proxy_ref.unwrap();
-    ns_proxy.uts_ns().set_domainname(addr, len, ctx)?;
+    if len > UtsName::MAX_FIELD_BYTES {
+        return_errno_with_message!(Errno::EINVAL, "the UTS name is too long");
+    }
```
Comment (lrh2000) [inline kernel/src/syscall/setdomainname.rs]: I don't think we really need to repeat the code? At least for now, we can still do:
```rust
let new_domain_name = UtsField::read_from(addr, len, ctx)?;

let ns_proxy_ref = ctx.thread_local.borrow_ns_proxy();
let ns_proxy = ns_proxy_ref.unwrap();
ns_proxy
    .uts_ns()
    .set_domainname(new_domain_name, ctx.posix_thread)?;

Ok(SyscallReturn::Return(0))
```

Also rearrange the code block more reasonably.

File: kernel/src/fs/fs_impls/procfs/sys/kernel/tainted.rs [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+
+    fn write_at(&self, _offset: usize, reader: &mut VmReader) -> Result<usize> {
+        let current_thread = current_thread!();
+        let posix_thread = current_thread.as_posix_thread().unwrap();
+
+        lsm_hooks::on_capable(lsm_hooks::CapableContext::new(
+            UserNamespace::get_init_singleton().as_ref(),
+            posix_thread,
+            CapSet::SYS_ADMIN,
+        ))?;
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/tainted.rs]:
```suggestion
        lsm_hooks::on_capable(lsm_hooks::CapableContext::new(
            UserNamespace::get_init_singleton().as_ref(),
            current_thread!().as_posix_thread().unwrap(),
            CapSet::SYS_ADMIN,
        ))?;
```

File: kernel/src/net/uts_ns.rs (hunk 18 above)
Comment (lrh2000) [reply kernel/src/net/uts_ns.rs]: Thanks!

**Review by lrh2000 — COMMENTED** [review]
Thanks for the updates! I was wondering how the APIs could be designed better. So far, these seem to be the best I can think of:

**Comment by bet4it** [comment]
Thanks a lot for the detailed review and for spending the time to think through the API design. I agree that introducing `UtsField` makes the code much clearer and easier to reason about. I have updated the patch following your suggested direction: the UTS field invariant is now encapsulated in `UtsField`, the procfs write path is simplified using your proposed structure, and the syscall paths now use `UtsField::read_from` instead of duplicating the parsing logic. I also inlined the `tainted` capability check as suggested.

This separation does look better to me: procfs keeps the procfs-specific write semantics, syscall keeps the strict userspace read behavior, and `UtsNamespace` only receives an already validated UTS field. Thanks again for the helpful suggestions.

File: kernel/src/net/uts_ns.rs [hunk 24]
```diff
@@ -206,6 +179,71 @@ impl UtsName {
...
+    /// Reads a UTS field from user space.
+    pub fn read_from(addr: Vaddr, len: usize, ctx: &Context) -> Result<Self> {
+        if len > Self::MAX_BYTES {
+            return_errno_with_message!(Errno::EINVAL, "the UTS name is too long");
+        }
+
+        let user_space = ctx.user_space();
+        let mut reader = user_space.reader(addr, len)?;
+        let mut field = [0u8; Self::MAX_BYTES_WITH_NUL];
+        reader.read_fallible(&mut VmWriter::from(&mut field[..len]))?;
```
Comment (lrh2000) [inline kernel/src/net/uts_ns.rs]: Let's keep the original code and comments.

```c
    // UTS fields represent C strings, which must be nul-terminated.
    // Therefore, the user-provided buffer length cannot exceed `UTS_FIELD_LEN - 1`
    // to ensure space for the terminating nul byte.
```

```c

    // Partial reads are acceptable,
    // but an error is returned if no bytes can be read successfully.
    if let Err((err, 0)) = reader.read_fallible(&mut VmWriter::from(buffer.as_mut_slice())) {
```

procfs rejects short writes. `sethostname` should follow a different convention. Don't change the existing code and/or existing comments for no good reason.

File: kernel/src/net/uts_ns.rs [hunk 25]
```diff
@@ -206,6 +179,71 @@ impl UtsName {
...
+
+        Ok(Self::from_bytes_until_nul(&field))
+    }
+
+    /// Returns the UTS field as a C string.
+    #[expect(clippy::wrong_self_convention, reason = "keep the UtsField API names")]
```
Comment (lrh2000) [inline kernel/src/net/uts_ns.rs]: Emmm.. It seems that we should use `as_`?

I am not sure why `CStr::to_bytes` use `to_`. I just copy the names here. But maybe we need to respect the clippy lint instead.

Renaming the method name seems to be a better idea than suppressing the lint in most cases.

File: kernel/src/net/uts_ns.rs [hunk 26]
```diff
@@ -206,6 +179,71 @@ impl UtsName {
...
+
+impl UtsField {
+    /// The maximum byte length of a UTS field, excluding the trailing nul.
+    pub const MAX_BYTES: usize = 64;
+
+    /// The storage byte length of a UTS field, including the trailing nul.
+    pub const MAX_BYTES_WITH_NUL: usize = Self::MAX_BYTES + 1;
+
+    /// Creates a UTS field from bytes, stopping at the first nul byte.
+    pub fn from_bytes_until_nul(bytes: &[u8]) -> Self {
```
Comment (lrh2000) [inline kernel/src/net/uts_ns.rs:216]: Add a paragraph to describe the truncation behavior in doc comments.

File: kernel/src/net/uts_ns.rs [hunk 27]
```diff
@@ -72,80 +71,54 @@ impl UtsNamespace {
...
+    pub fn set_hostname(&self, new_host_name: UtsField, posix_thread: &PosixThread) -> Result<()> {
+        self.check_set_permission(posix_thread)?;
+        self.set_hostname_field(new_host_name);
         Ok(())
     }
 
     /// Sets a new domain name for the UTS namespace.
     ///
     /// This method will fail with `EPERM` if the caller does not have the SYS_ADMIN capability
     /// in the owner user namespace.
```
Comment (lrh2000) [inline kernel/src/net/uts_ns.rs]: "the caller" -> "the POSIX thread". given this method no longer accepts `Context` .

**Review by lrh2000 — COMMENTED** [review]
Thanks for your updates. Just some minor comments.

File: kernel/src/net/uts_ns.rs
Comment (lrh2000) [inline kernel/src/net/uts_ns.rs:162]: Although not introducing in this PR, could you please update this line to:
```rust
assert!(VERSION.len() <= UtsField::MAX_BYTES);
```
?

(I think we don't need to use `const_assert` here...)

File: kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs [hunk 28]
```diff
@@ -0,0 +1,172 @@
...
+
+fn write_current_uts_name(
+    offset: usize,
+    reader: &mut VmReader,
+    get_value: fn(&UtsName) -> &UtsField,
+    set_value: fn(&UtsNamespace, UtsField, &PosixThread) -> Result<()>,
+) -> Result<usize> {
+    let len = reader.remain();
+    if offset >= UtsField::MAX_BYTES || len == 0 {
+        reader.skip(len);
```
Comment (lrh2000) [inline kernel/src/fs/fs_impls/procfs/sys/kernel/uts.rs:133]: I think we can keep the suggested comment (https://github.com/asterinas/asterinas/pull/3522#discussion_r3549484153):
```c
        // Overflowed bytes are consumed but ignored.
```
in line 133 and line 170. This makes the intention clearer.

Or do you have a reason not to do so / have a better version than the suggestion?

**Review by lrh2000 — APPROVED** [review]
LGTM except two minor nits. Thanks for your contribution!

**Comment by lrh2000** [comment]
Oh, obviously there are some CI issues :(

**Comment by bet4it** [comment]
I rebased onto the latest `origin/main`, which includes the recent #3492 changes, and tried to enable `fsconfig03` since the previous blocker, missing `/proc/sys/kernel/tainted`, has been fixed. However, CI showed that our current `fsconfig` behavior still has a compatibility issue with `fsconfig03`😅

So I will keep `fsconfig03` disabled in this PR for now, and follow up with a separate PR to fix the `fsconfig` behavior and enable the test after this PR is merged.

**Review by cqs21 — APPROVED** [review]
LGTM! Thanks for your contribution!
